### PR TITLE
Feature/patro23 48 mediat r as command query bus

### PR DESCRIPTION
**Pull request by AlanP22**
PATRO23-48: Added command and query bus implementation.

### CODE REVIEW

File: src/shared/abstractions/Commands/CommandBus.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+using System;
```
Comment (mateuszmuszynski) [inline src/shared/abstractions/Commands/CommandBus.cs:1]: Abstractions project shouldn't contain actual implementation of command bus. Move `CommandBus` implementation to Infrastructure project

File: src/shared/abstractions/Commands/ICommandBus.cs [hunk 2]
```diff
@@ -11,6 +11,6 @@ public interface ICommandBus
 	/// <typeparam name="TCommand">Type of command.</typeparam>
 	/// <param name="command">Command to send.</param>
 	/// <returns>Task that represents asynchronous operation.</returns>
-	Task Send<TCommand>(TCommand command)
+	Task<object> Send<TCommand>(TCommand command)
```
Comment (mateuszmuszynski) [inline src/shared/abstractions/Commands/ICommandBus.cs:14]: Send method shouldn't return any data

**Review by mateuszmuszynski — CHANGES_REQUESTED** [review]

File: src/shared/abstractions/Commands/ICommandBus.cs (hunk 2 above)
Comment (AlanP22) [reply src/shared/abstractions/Commands/ICommandBus.cs:14]: @mateuszmuszynski Can it be asynchronous if it doesn't return any data? Is it even necessary?

Comment (mateuszmuszynski) [reply src/shared/abstractions/Commands/ICommandBus.cs]: Let's leave it as asynchronous.

Comment (mateuszmuszynski) [reply src/shared/abstractions/Commands/ICommandBus.cs]: Please also add an registration of Command/Query bus

Comment (AlanP22) [reply src/shared/abstractions/Commands/ICommandBus.cs]: Resolved in latest commits.

**Review by mateuszmuszynski — APPROVED** [review]

File: src/api/app/Program.cs [hunk 3]
```diff
@@ -40,6 +42,9 @@
 	xmlFiles.ForEach(xmlFile => options.IncludeXmlComments(xmlFile));
 });
 
+builder.Services.AddSingleton<CommandBus>();
+builder.Services.AddSingleton<QueryBus>();
```
Comment (IzabelaPalubicka) [inline src/api/app/Program.cs:46]: This should be added with abstraction <ICommandBus, CommandBus>, but automatic registration is on dev,  so you don't need to add it at all. Automatic registration is set by default as scoped, so maybe you have to change life time atribute like here [https://github.com/intive/patronage2023-dotnet/blob/dev/src/modules/example/infrastructure/Domain/EventHandlers/ExampleCreatedDomainEventHandler.cs](url)

**Review by IzabelaPalubicka — CHANGES_REQUESTED** [review]

File: src/api/app/Program.cs (hunk 3 above)
Comment (AlanP22) [reply src/api/app/Program.cs:51]: Added Lifetime scoped attributes in latest commit.

**Review by Kepeczsz — APPROVED** [review]

File: src/shared/infrastructure/Query/QueryBus.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		}
+
+		/// <inheritdoc/>
+		public async Task<TResponse> Query<TRequest, TResponse>(TRequest query)
+		{
+			var result = await this.mediator.Send(query!);
+
+			if (result == null)
+			{
+				throw new ArgumentNullException();
```
Comment (mateuszgiedzinski) [inline src/shared/infrastructure/Query/QueryBus.cs:30]: Which argument is null? Argument Exceptions are dedicated for method parameters only.  Better to use InvalidOperationException or InvalidDataException

**Review by mateuszgiedzinski — CHANGES_REQUESTED** [review]

File: src/shared/infrastructure/Query/QueryBus.cs (hunk 4 above)
Comment (mateuszgiedzinski) [reply src/shared/infrastructure/Query/QueryBus.cs:30]: Also you can change var to object?.

**Review by mateuszgiedzinski — APPROVED** [review]
Can be merged. There is a problem with namespaces because we should use file scoped namespaces  to have one less indentiation.  (namespace Intive.Patronage2023.Shared.Infrastructure.Commands;)
But because this is blocker for some task I can approve it conditionally

File: src/shared/infrastructure/Query/QueryBus.cs (hunk 4 above)
Comment (AlanP22) [reply src/shared/infrastructure/Query/QueryBus.cs]: Done

**Review by IzabelaPalubicka — APPROVED** [review]

**Pull request merged**